crates/iceberg-catalog/src/service/storage/gcs/sts.rs

```diff
@@ -0,0 +1,226 @@
+use crate::service::storage::error::TableConfigError;
+use crate::service::storage::gcs::GcsServiceKey;
+use crate::service::storage::StoragePermissions;
+use google_cloud_auth::credentials::CredentialsFile;
+use iceberg_ext::configs::Location;
+use serde::{Deserialize, Serialize};
+use std::sync::OnceLock;
+use url::Url;
+
+static STS_CLIENT: OnceLock<reqwest::Client> = OnceLock::new();
+static STS_URL: OnceLock<Url> = OnceLock::new();
+const STS_URL_STR: &str = "https://sts.googleapis.com/v1/token";
+const GOOGLE_CLOUD_PLATFORM_SCOPE: &str = "https://www.googleapis.com/auth/cloud-platform";
+
+pub(crate) async fn downscope(
+    cred: impl Into<CredentialsFile>,
+    bucket: &str,
+    table_location: Location,
+    storage_permissions: StoragePermissions,
+) -> Result<STSResponse, TableConfigError> {
+    let client = STS_CLIENT.get_or_init(reqwest::Client::new);
+    let sts_url = STS_URL
+        .get_or_init(|| {
+            STS_URL_STR
+                .parse::<Url>()
+                .expect("failed to parse a constant to a url")
+        })
+        .clone();
+
+    let c =
+        google_cloud_auth::project::Config::default().with_scopes(&[GOOGLE_CLOUD_PLATFORM_SCOPE]);
+    let source = google_cloud_auth::project::create_token_source_from_credentials(&cred.into(), &c)
+        .await
+        .map_err(|e| {
+            tracing::error!(
+                "Failed to create gcp token source from credentials: {:?}",
+                e
+            );
+            TableConfigError::FailedDependency(
+                "Failed to create gcp token source from credentials".to_string(),
+            )
+        })?;
+    let token = source.token().await.map_err(|e| {
+        tracing::error!("Failed to get token from token source: {:?}", e);
+        TableConfigError::FailedDependency("Failed to get gcp token from token source".to_string())
+    })?;
+
+    client
+        .post(sts_url)
+        .header("Content-Type", "application/json")
+        .json(&STSRequest::from_token_and_options(
+            &token.access_token,
+            &Options::from_location_and_permissions(bucket, &table_location, storage_permissions),
+        )?)
+        .send()
+        .await
+        .map_err(|e| {
+            tracing::error!("Failed to send downscoping request: {:?}", e);
+            TableConfigError::FailedDependency("Failed to send downscoping request".to_string())
+        })?
+        .json::<STSResponse>()
+        .await
+        .map_err(|e| {
+            tracing::error!("Failed to parse downscoping response: {:?}", e);
+            TableConfigError::StsError("Downscoping failed".to_string(), Some(Box::new(e)))
+        })
+}
+
+#[derive(Deserialize, veil::Redact)]
+pub(crate) struct STSResponse {
+    #[redact(partial)]
+    pub(crate) access_token: String,
+    pub(crate) expires_in: Option<usize>,
+    token_type: String,
```

Comment:
Should we introduce an enum, which we could also re-use in the request?

Reply:
we don't really use the token_type, we can remove it

Reply:
i think we should keep it, seems useful for debugging, i think it's not the same enum as in the request, in the request only a single string is valid anyways